source/ada/lsp-ada_handlers-file_readers.adb

```diff
@@ -27,6 +32,104 @@ package body LSP.Ada_Handlers.File_Readers is
 
    Me : constant Trace_Handle := Create ("ALS.FILE_READERS");
 
+   function Read_And_Convert_To_UTF8
+     (Filename : String; Charset : String)
+      return GNAT.Strings.String_Access;
+   --  Read the file content from Filename and convert it from the original
+   --  Charset to UTF-8.
+
+   ------------------------------
+   -- Read_And_Convert_To_UTF8 --
+   ------------------------------
+
+   function Read_And_Convert_To_UTF8
+     (Filename : String; Charset : String)
+      return GNAT.Strings.String_Access
+   is
+      Raw        : GNAT.Strings.String_Access;
+      Decoded    : GNAT.Strings.String_Access;
+   begin
+      --  Read the file (this call uses MMAP)
+      Raw := Create_From_UTF8 (Filename).Read_File;
+
+      if Raw = null then
+         return null;
+      end if;
+
+      --  Convert the file if it's not already encoded in utf-8
+
+      if Ada.Characters.Handling.To_Lower (Charset) = "utf-8" then
+         Decoded := Raw;
+      else
+         declare
+            State        : constant Iconv_T := Iconv_Open (UTF8, Charset);
+            Outbuf       : Byte_Sequence (1 .. 4096);
+            Input_Index  : Positive := Raw'First;
+            Conv_Result  : Iconv_Result := Full_Buffer;
+            Output_Index : Positive;
+         begin
+            while Conv_Result = Full_Buffer loop
+               Output_Index := 1;
+               Iconv (State        => State,
+                      Inbuf        => Raw.all,
+                      Input_Index  => Input_Index,
+                      Outbuf       => Outbuf,
+                      Output_Index => Output_Index,
+                      Result       => Conv_Result);
+
+               --  Append the converted contents
+               if Decoded /= null then
```

Comment:
Decoded is always null here.

I recommend to use VSS.Strings.Converters here instead of Iconv. We can add necessary encodings to VSS, it is not complicated; however, VSS's decoders is able to recover after conversion error, which may be important for us - it save as much information as possible, instead of dropping everything after first failure.

Reply:
Indeed, It's Raw that I need to free, thanks.

For the VSS stuff, if you know how to proceed feel free to make the modifications! I can create a separate ticket for that. I just want to fix the testsuite meanwhile.